src/common/session/session-entity.js

```diff
@@ -0,0 +1,219 @@
+    // if the initial read is empty, it allows us to set a "fresh" "new" session immediately.
+    // the local timers are used after the session is running to "expire" the session, allowing for pausing timers etc.
+    // the timestamps stored in the storage API can be checked at initial run, and when the page is restored, otherwise we lean
+    // on the local timers to expire the session
+    const initialRead = this.read()
+
+    // the set-up of the timer used to expire the session "naturally" at a certain time
+    // this gets ignored if the value is falsy, allowing for session entities that do not expire
+    this.expiresMs = expiresMs
+    if (expiresMs) {
+      this.expiresAt = initialRead?.expiresAt || this.getFutureTimestamp(expiresMs)
+      this.expiresTimer = new Timer({
+        onEnd: () => this.reset()
+      }, expiresMs)
+      // this.expiresTimer = new Timer(() => this.reset(), this.expiresAt - Date.now())
+    } else {
+      this.expiresAt = Infinity
+    }
+
+    // the set-up of the timer used to expire the session due to "inactivity" at a certain time
+    // this gets ignored if the value is falsy, allowing for session entities that do not expire
+    // this gets "refreshed" when "activity" is observed
+    this.inactiveMs = inactiveMs
+    if (inactiveMs) {
+      this.inactiveAt = initialRead?.inactiveAt || this.getFutureTimestamp(inactiveMs)
+      this.inactiveTimer = new InteractionTimer({
+        onEnd: () => this.reset(),
+        onRefresh: () => this.refresh(),
+        ee: ee.get(agentIdentifier),
+        refreshEvents: ['click', 'keydown', 'scroll'],
+        abortController: this.abortController
+      }, inactiveMs)
+    } else {
+      this.inactiveAt = Infinity
+    }
+
+    // The fact that the session is "new" or pre-existing is used in some places in the agent.  Session Replay and Trace
+    // can use this info to inform whether to trust a new sampling decision vs continue a previous tracking effort.
+    this.isNew = !Object.keys(initialRead).length
+    // if its a "new" session, we write to storage API with the default values.  These values ,ay change over the lifespan of the agent run.
+    if (this.isNew) this.write({ value, sessionReplayActive: false, sessionTraceActive: false, inactiveAt: this.inactiveAt, expiresAt: this.expiresAt }, true)
+    else {
+      Object.keys(initialRead).forEach(k => {
+        this[k] = initialRead[k]
+      })
+    }
+
+    this.initialized = true
+  }
+
+  // This is the actual key appended to the storage API
+  get lookupKey () {
+    return `${PREFIX}_${this.key}`
+  }
+
+  /**
+   * Fetch the stored values from the storage API tied to this entity
+   * @returns {Object}
+   */
+  read () {
+    try {
+      const val = this.storage.get(this.lookupKey)
+      if (!val) return {}
+      // TODO - decompression would need to happen here if we decide to do it
+      const obj = typeof val === 'string' ? JSON.parse(val) : val
```

Comment:
This further enforces that really it's `localStorage` & `FP cookies`, and that `localMemory` isn't really a "storage API option".
The first two only ever store strings (will forcefully convert), _and_ "local memory" aka just-an-object is no different from how the current state of data lives in an initialized agent.

Following my other comment about storing all the key-values inside one central object, I'm thinking you could set/update/remove value in this intermediary, and only ever read (on vis to visible) and write (on vis to hidden) _once_, akin to React's virtual DOM.

Reply:
Let me know if I'm confusing this with the storage medium for which customer-defined data would persist. I assume it's through this interface.

Reply:
i'm not really following what you are saying here